lib/kubernetestoken/token_validator.go

```diff
@@ -0,0 +1,133 @@
+	// method.
+	client kubernetes.Interface
+}
+
+// getClient allows the lazy initialisation of the Kubernetes client
+func (v *Validator) getClient() (kubernetes.Interface, error) {
+	v.mu.Lock()
+	defer v.mu.Unlock()
+	if v.client != nil {
+		return v.client, nil
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes config")
+	}
+	// Intentionally use context.Background() here since this actually controls
+	// cache functionality.
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, trace.WrapWithMessage(err, "failed to initialize in-cluster Kubernetes client")
+	}
+
+	v.client = client
+	return client, nil
+}
+
+// Validate uses the Kubernetes TokenReview API to validate a token and return its UserInfo
+func (v *Validator) Validate(ctx context.Context, token string) (*v1.UserInfo, error) {
+	client, err := v.getClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	review := &v1.TokenReview{
+		Spec: v1.TokenReviewSpec{
+			Token: token,
+		},
+	}
+	options := metav1.CreateOptions{}
+
+	reviewResult, err := client.AuthenticationV1().TokenReviews().Create(ctx, review, options)
+	if err != nil {
+		return nil, trace.WrapWithMessage(err, "error during the Kubernetes TokenReview")
+	}
+
+	if !reviewResult.Status.Authenticated {
+		return nil, trace.AccessDenied("kubernetes failed to validate token: %s", reviewResult.Status.Error)
+	}
+
+	// Legacy tokens are long-lived and not bound to pods. We should not accept them if the cluster supports
+	// bound tokens. Bound token support is GA since 1.20 and volume projection is beta since 1.21.
+	// We can expect any 1.21+ cluster to use bound tokens.
+	kubeVersion, err := client.Discovery().ServerVersion()
+	if err != nil {
+		return nil, trace.WrapWithMessage(err, "error during the kubernetes version check")
+	}
+
+	boundTokenSupport, err := kubernetesSupportsBoundTokens(kubeVersion)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// We know if the token is bound to a pod if its name is in the Extra userInfo.
+	// If the token is not bound while Kubernetes supports bound tokens we abort.
+	if _, ok := reviewResult.Status.User.Extra[ExtraDataPodNameField]; !ok && boundTokenSupport {
+		return nil, trace.BadParameter("legacy tokens are not suppoprted for kubernetes versions (%s) supporting bound tokens", kubeVersion.GitVersion)
+	}
+
+	// Check the Username is a service account.
+	// A user token would not match rules anyway, but we can produce a more relevant error message here.
+	if !strings.HasPrefix(reviewResult.Status.User.Username, ServiceAccountNamePrefix) {
+		return nil, trace.BadParameter("token user is not a service account: %s", reviewResult.Status.User.Username)
+	}
+
+	return &reviewResult.Status.User, nil
+}
+
+func kubernetesSupportsBoundTokens(info *version.Info) (bool, error) {
```

Comment:
Ugh, I'm not sure and it's not clear how to validate this either, reproducing this scenario will be a pain and the result might depend on the deployment method. The capability is in the kubelet but it might require a feature gate flag.

It's a bit sad but I can bump the restriction to 1.22.

Reply:
I think bumping to 1.22 is probably the "safest" option here that is also the simplest :')

Reply:
done, with a comment

Reply:
Perfect !